### ISSUE TITLE
spearman_corrcoef and kendall_rank_corrcoef silently downcast float64 to float32

### ISSUE THREAD

**Issue by VenishPaneliya** [Lightning-AI/torchmetrics#3465]
## 🐛 Bug

`spearman_corrcoef` and `kendall_rank_corrcoef` silently return `float32` for `float64` inputs. Every other regression metric preserves the input dtype, and `kendall_rank_corrcoef` itself preserves it for `variant="c"` but not for `"a"` or `"b"`, so this looks unintended rather than deliberate.

### To Reproduce

```python
import torch
from torchmetrics.functional import pearson_corrcoef, spearman_corrcoef, kendall_rank_corrcoef

preds = torch.rand(100, dtype=torch.float64)
target = torch.rand(100, dtype=torch.float64)

print(pearson_corrcoef(preds, target).dtype)       # torch.float64
print(spearman_corrcoef(preds, target).dtype)      # torch.float32  <-- 
print(kendall_rank_corrcoef(preds, target).dtype)  # torch.float32  <-- 
```

The class-based `SpearmanCorrCoef` and `KendallRankCorrCoef` return `float32` as well.

`kendall_rank_corrcoef` is inconsistent across its own variants:

| `variant` | output dtype for `float64` input |
|---|---|
| `"a"` | `float32` |
| `"b"` (default) | `float32` |
| `"c"` | `float64` |

Surveying the functional regression metrics with `float64` inputs, these two are the only ones that downcast:

| metric | float64 in -> out |
|---|---|
| `mean_absolute_error`, `mean_squared_error`, `mean_absolute_percentage_error`, `symmetric_mean_absolute_percentage_error`, `mean_squared_log_error`, `weighted_mean_absolute_percentage_error`, `r2_score`, `explained_variance`, `pearson_corrcoef`, `concordance_corrcoef`, `log_cosh_error`, `relative_squared_error`, `tweedie_deviance_score`, `minkowski_distance`, `normalized_root_mean_squared_error` | `float64` |
| **`spearman_corrcoef`** | **`float32`** |
| **`kendall_rank_corrcoef`** | **`float32`** |

### Why it matters beyond the dtype label

Spearman is rank-based, so two identical strictly-increasing vectors must give exactly `1.0` regardless of the spacing between values. When the differences are below `float32` resolution the result drifts:

```python
import torch
from torchmetrics.functional import spearman_corrcoef, pearson_corrcoef

a = torch.tensor([1.0, 1.0 + 1e-10, 1.0 + 2e-10, 1.0 + 3e-10], dtype=torch.float64)

print(spearman_corrcoef(a, a).item())  # 0.9999992251396179
print(pearson_corrcoef(a, a).item())   # 0.9999999999999999
```

Users who deliberately opt into `float64` — common when correlating values with a small dynamic range — silently lose that precision.

### Cause

Both come from an integer intermediate being divided, which falls back to the **global default dtype** (`float32`) rather than the input dtype.

**`spearman_corrcoef`** — `_rank_data` in `functional/regression/spearman.py`:

```python
sum_ranks = torch.zeros_like(uniq, dtype=torch.int32)
sum_ranks.scatter_add_(0, inv, rank.to(torch.int32))
mean_ranks = sum_ranks / counts   # int32 / int64 -> float32, regardless of data.dtype
return mean_ranks[inv]
```

The ranks therefore come back `float32` and everything downstream inherits it.

**`kendall_rank_corrcoef`** — `_get_metric_metadata` in `functional/regression/kendall.py`:

```python
n_total = torch.tensor(preds.shape[0], device=preds.device)   # int64, no dtype from preds
```

The concordant/discordant/tie counts are `int64` too, so in `_calculate_tau` the divisions for variants `"a"` and `"b"` produce `float32`. Variant `"c"` escapes it only because it builds `preds_unique` with `dtype=preds.dtype`, which promotes the expression.

### Expected behavior

The output dtype should follow the input, as the other regression metrics do — `float64` in, `float64` out. For `float16`/`bfloat16` inputs, promoting the rank/count arithmetic to `float32` is still appropriate, since ranks are integers that need more mantissa than `float16` offers for larger inputs.

### Environment

- TorchMetrics: `master` (commit 37a80809)
- PyTorch: 2.13.0+cpu, Python 3.13, Windows

I have a fix and regression tests ready and am happy to open a PR if this is welcome.

**Comment by VenishPaneliya** [comment]
Correction to my own report, after building the fix and measuring properly.

**The `0.9999992` example above is mostly not caused by the downcast.** I attributed it to `float32` and that was wrong. The dominant term is the `eps=1e-6` added to the denominator in `_spearman_corrcoef_compute`:

```python
corrcoef = cov / (preds_std * target_std + eps)
```

Isolating it on exact integer ranks, where the dtype plays no part:

```
_spearman_corrcoef_compute(a, a)             # eps=1e-6 (default) -> 0.9999992000006399
_spearman_corrcoef_compute(a, a, eps=0.0)    #                    -> 0.9999999999999998
```

With the dtype fix applied but `eps` untouched, the example still returns `0.9999992000006399` — it only gains float64 *representation*, not accuracy. So `eps` biasing every result away from an exact ±1 is a **separate** issue from the downcast. I'm deliberately not touching it, since changing `eps` would shift results for every existing user and is a call for you rather than me. Happy to open a separate issue for it if you'd like it tracked.

The dtype problem itself is real and stands on its own — `float64` in should give `float64` out, as the other regression metrics do, and `kendall_rank_corrcoef`'s variant `"c"` already does.

**Second finding, at the class level.** `SpearmanCorrCoef.update` casts explicitly:

```python
self.preds.append(preds.to(self.dtype))     # self.dtype is the module dtype, float32 by default
self.target.append(target.to(self.dtype))
```

`KendallRankCorrCoef.update` has no such cast, which is why the class-based Kendall preserves `float64` while the class-based Spearman cannot, regardless of what the functional path does. I don't know whether that cast is deliberate — halving the stored footprint would be a reasonable motive, given the metric already warns about buffering all predictions — so I've left it alone and scoped my PR to the functional path. Let me know if you'd like the class-level behaviour aligned too and I'll add it.
